### PR TITLE
add `CalculateTurnPlayerValuesEvent`

### PR DESCRIPTION
**Pull request by Nolij**
This PR adds an event called `CalculateTurnPlayerValuesEvent` which allows hooking the mouse sensitivity values and cinematic camera values from the game options when running `turnPlayer`, which is responsible for rotating the camera in response to mouse movement.

### CODE REVIEW

**Comment by ChampionAsh5357** [comment]
What benefit does this event provide over manually updating the values yourself in the options? They are public and mutable.

**Comment by Nolij** [comment]
> What benefit does this event provide over manually updating the values yourself in the options? They are public and mutable.

This allows hooking instead of direct modifications. I intend to use this event to make the NeoForge implementation of [Zume](https://github.com/Nolij/Zume) (a mod I made) mixinless. I use mixins so that I don't need to manually maintain the values in options. The alternative would be setting the value every frame or messing with keybinds or something.

**Comment by ChampionAsh5357** [comment]
> This allows hooking instead of direct modifications. I intend to use this event to make the NeoForge implementation of [Zume](https://github.com/Nolij/Zume) (a mod I made) mixinless. I use mixins so that I don't need to manually maintain the values in options. The alternative would be setting the value every frame or messing with keybinds or something.

The event isn't cancelable nor has a result. People lower down the chain will have their values take priority. The getters are already public and mutable without the event. The event will also be called every frame for a value that can be set statically, which is something to be avoided when possible.

The only thing I can see here is that it prevents the player from overriding the decisions, which I feel shouldn't be a concern. If they are messing with the value, then there will be unintended issues anyway.

If someone else on the team has a differing opinion, I'm all ears.

**Comment by Nolij** [comment]
> > This allows hooking instead of direct modifications. I intend to use this event to make the NeoForge implementation of [Zume](https://github.com/Nolij/Zume) (a mod I made) mixinless. I use mixins so that I don't need to manually maintain the values in options. The alternative would be setting the value every frame or messing with keybinds or something.
> 
> The event isn't cancelable nor has a result. People lower down the chain will have their values take priority. The getters are already public and mutable without the event. The event will also be called every frame for a value that can be set statically, which is something to be avoided when possible.
> 
> The only thing I can see here is that it prevents the player from overriding the decisions, which I feel shouldn't be a concern. If they are messing with the value, then there will be unintended issues anyway.
> 
> If someone else on the team has a differing opinion, I'm all ears.

The same issue with priority and recalculating every frame also applies to the `ComputeFov` event (which I also use). I figured this was an acceptable use of events given the existence of that one.

Also, in the case of mouse sensitivity, doing things the way you suggest could lead to edge cases where the user's configured mouse sensitivity is overwritten (and saved, persisting even after a problematic mod is removed) due to a bug. My hooking approach avoids this happening unintentionally.

If this event is not added, then I'll just continue using mixins to achieve the same effect. I think this event is not unprecedented though (given the existence of other events such as `ComputeFov`) and would be of benefit to users and mod devs to add.

**Comment by embeddedt** [comment]
As @Nolij mentioned, directly modifying the options has the side effect of allowing the change to persist indefinitely if the modification is not reverted before the options are saved. I don't have an issue with this event personally.

**Comment by ChampionAsh5357** [comment]
> The same issue with priority and recalculating every frame also applies to the `ComputeFov` event (which I also use). I figured this was an acceptable use of events given the existence of that one.

It is. My general opinion is just to avoid events that execute every frame or tick unless there is no better option.

> Also, in the case of mouse sensitivity, doing things the way you suggest could lead to edge cases where the user's configured mouse sensitivity is overwritten (and saved, persisting even after a problematic mod is removed) due to a bug. My hooking approach avoids this happening unintentionally.

Fair point. I did not think about session edge cases. My apologies.

I don't see anything this can be combined with, so the event seems reasonable to me. If you could provide a test mod as well, that would be great.

**Review by embeddedt — DISMISSED** [review]

**Comment by Nolij** [comment]
> > The same issue with priority and recalculating every frame also applies to the `ComputeFov` event (which I also use). I figured this was an acceptable use of events given the existence of that one.
> 
> It is. My general opinion is just to avoid events that execute every frame or tick unless there is no better option.
> 
> > Also, in the case of mouse sensitivity, doing things the way you suggest could lead to edge cases where the user's configured mouse sensitivity is overwritten (and saved, persisting even after a problematic mod is removed) due to a bug. My hooking approach avoids this happening unintentionally.
> 
> Fair point. I did not think about session edge cases. My apologies.
> 
> I don't see anything this can be combined with, so the event seems reasonable to me. If you could provide a test mod as well, that would be great.

Not sure if this is what you mean by "test mod" but [the `experiments/mixinless-neozume` branch](https://github.com/Nolij/Zume/tree/experiments/mixinless-neozume) of Zume (the mod I mentioned earlier) utilizes this event.

**Review by Technici4n — COMMENTED** [review]
Looks good. The naming is a bit convoluted but for a niche event I don't really mind.

**Review by Technici4n — COMMENTED** [review]
Maybe something like `CalculatePlayerTurnEvent`?

File: patches/net/minecraft/client/MouseHandler.java.patch [hunk 1]
```diff
@@ -65,6 +65,22 @@
                      if (this.minecraft.player.isSpectator()) {
                          if (this.minecraft.gui.getSpectatorGui().isMenuActive()) {
                              this.minecraft.gui.getSpectatorGui().onMouseScrolled(-k);
+@@ -235,12 +_,13 @@
+         double d1 = d0 - this.lastMouseEventTime;
+         this.lastMouseEventTime = d0;
+         if (this.isMouseGrabbed() && this.minecraft.isWindowActive()) {
+-            double d4 = this.minecraft.options.sensitivity().get() * 0.6F + 0.2F;
++            final net.neoforged.neoforge.client.event.CalculateTurnPlayerValuesEvent event = net.neoforged.neoforge.client.ClientHooks.getTurnPlayerValues(this.minecraft.options.sensitivity().get(), this.minecraft.options.smoothCamera);
```
Comment (Technici4n) [inline patches/net/minecraft/client/MouseHandler.java.patch]: You can use `var` to make the line a lot shorter.

File: src/main/java/net/neoforged/neoforge/client/event/CalculateTurnPlayerValuesEvent.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    private boolean cinematicCameraEnabled;
+
+    @ApiStatus.Internal
+    public CalculateTurnPlayerValuesEvent(double mouseSensitivity, boolean cinematicCameraEnabled) {
+        setMouseSensitivity(mouseSensitivity);
+        setCinematicCameraEnabled(cinematicCameraEnabled);
+    }
+
+    /**
+     * @return the raw {@linkplain Options#sensitivity() mouse sensitivity} value
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/client/event/CalculateTurnPlayerValuesEvent.java]: With only `@return`, the method won't get a summary if one ever renders the javadoc. It would be preferable to state `Returns <content>.`. (Or use the `{@return ...}` tag but I would advise against it because intellij will not render the nested tag correctly).

File: patches/net/minecraft/client/MouseHandler.java.patch (hunk 1 above)
Comment (Nolij) [reply patches/net/minecraft/client/MouseHandler.java.patch]: Assumed I wasnt allowed to use `var` in patches based on the rest of the file. Will do.

File: src/main/java/net/neoforged/neoforge/client/event/CalculateTurnPlayerValuesEvent.java (hunk 2 above)
Comment (Nolij) [reply src/main/java/net/neoforged/neoforge/client/event/CalculateTurnPlayerValuesEvent.java]: Didn't know that. I'll do "Returns ..."

**Comment by Nolij** [comment]
> Maybe something like `CalculatePlayerTurnEvent`?

This name works for me.

**Comment by Nolij** [comment]
Also updated the experiment branch on Zume for testing purpose.

**Review by embeddedt — APPROVED** [review]
